core/variant_call.cpp

```diff
@@ -35,1058 +35,759 @@
+	arr.push_back(p_str);
+}
+
+template <class... P>
+_FORCE_INLINE_ void sarray_add_str(Vector<String> &arr, const String &p_str, P... p_args) {
+	arr.push_back(p_str);
+	sarray_add_str(arr, p_args...);
+}
+
+template <class... P>
+_FORCE_INLINE_ Vector<String> sarray(P... p_args) {
+	Vector<String> arr;
+	sarray_add_str(arr, p_args...);
+	return arr;
+}
+
 typedef void (*VariantFunc)(Variant &r_ret, Variant &p_self, const Variant **p_args);
 typedef void (*VariantConstructFunc)(Variant &r_ret, const Variant **p_args);
 
 struct _VariantCall {
-	static void Vector3_dot(Variant &r_ret, Variant &p_self, const Variant **p_args) {
-		r_ret = reinterpret_cast<Vector3 *>(p_self._data._mem)->dot(*reinterpret_cast<const Vector3 *>(p_args[0]->_data._mem));
-	}
-
-	struct FuncData {
-		int arg_count;
-		Vector<Variant> default_args;
-		Vector<Variant::Type> arg_types;
-		Vector<StringName> arg_names;
-		Variant::Type return_type;
+	template <class T, class... P>
+	class InternalMethod : public Variant::InternalMethod {
+	public:
+		void (T::*method)(P...);
+		Vector<Variant> default_values;
+#ifdef DEBUG_ENABLED
+		Vector<String> argument_names;
+#endif
 
-		bool _const;
-		bool returns;
+		virtual int get_argument_count() const {
+			return sizeof...(P);
+		}
+		virtual Variant::Type get_argument_type(int p_arg) const {
+			return call_get_argument_type<P...>(p_arg);
+		}
+#ifdef DEBUG_ENABLED
+		virtual String get_argument_name(int p_arg) const {
+			ERR_FAIL_INDEX_V(p_arg, argument_names.size(), String());
+			return argument_names[p_arg];
+		}
+#endif
+		virtual Vector<Variant> get_default_arguments() const {
+			return default_values;
+		}
 
-		VariantFunc func;
+		virtual Variant::Type get_return_type() const {
+			return Variant::NIL;
+		}
+		virtual uint32_t get_flags() const {
+			return 0;
+		}
 
-		_FORCE_INLINE_ bool verify_arguments(const Variant **p_args, Callable::CallError &r_error) {
-			if (arg_count == 0) {
-				return true;
+		virtual void call(Variant *base, const Variant **p_args, int p_argcount, Variant &r_ret, Callable::CallError &r_error) {
+			const Variant **args = p_args;
+#ifdef DEBUG_ENABLED
+			if ((size_t)p_argcount > sizeof...(P)) {
+				r_error.error = Callable::CallError::CALL_ERROR_TOO_MANY_ARGUMENTS;
+				r_error.argument = sizeof...(P);
+				return;
 			}
+#endif
+			if ((size_t)p_argcount < sizeof...(P)) {
+				size_t missing = sizeof...(P) - (size_t)p_argcount;
+				if (missing <= (size_t)default_values.size()) {
+					args = (const Variant **)alloca(sizeof...(P) * sizeof(const Variant *));
+					// GCC fails to see that `sizeof...(P)` cannot be 0 here given the previous
+					// conditions, so it raises a warning on the potential use of `i < 0` as the
+					// execution condition.
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wtype-limits"
+#endif
+					for (size_t i = 0; i < sizeof...(P); i++) {
```

Comment:
This might be worth a bug report to upstream GCC IMO, if anyone is motivated to write a minimal reproduction project (basically some code that uses variadic templates with a parameter pack that can be of size 0, but which has a check for its size before attempting to use it in a comparison with a `size_t`).